Add Python 3.7 to Travis

asmeurer opened the pull request:
This also moves all the 3.6-only builds like optional dependency, docs, and
slow tests from 3.6 to 3.7.

Note: I'm not 100% sure this will work. I think all the conda packages from the optional dependency build should be available for 3.7, but I'm not positive, especially for Sage. 

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

asmeurer committed:
Add Python 3.7 to Travis

This also moves all the 3.6-only builds like optional dependency, docs, and
slow tests from 3.6 to 3.7.

asmeurer committed:
Python 3.7 requires dist: xenial, sudo: true to work in Travis

See https://github.com/travis-ci/travis-ci/issues/9815.

asmeurer commented:
Unfortunately, this adds to the total number of builds.

The good news is that once we get this next release out (1.4), I think we can safely drop Python 3.4 support. See https://github.com/sympy/sympy/wiki/Python-version-support-policy and https://devguide.python.org/#status-of-python-branches. So it should only be temporary.

oscarbenjamin commented:
Tests for tensorflow are failing. Tensorflow isn't available for Python 3.7 yet. Maybe those tests should be skipped on Python 3.7 for now.

asmeurer committed:
Test not installing tensorflow in Python 3.7 on Travis

This is to see if any other packages have issues with Python 3.7.

asmeurer commented:
Looks like tensorflow itself doesn't support 3.7 yet (https://github.com/conda-forge/tensorflow-feedstock/issues/65). I am testing if any other packages are also lagging, but I think we should keep the optional dependency build on 3.6 for now.

asmeurer commented:
Looks like our TEST_ASCII is no longer necessary in Python 3.7 due to https://docs.python.org/3/whatsnew/3.7.html#pep-538-legacy-c-locale-coercion. Unicode characters are now supported even if the locale doesn't support them. I'm adding `PYTHONCOERCECLOCALE=0` to force this to have the pre-3.7 behavior, but once we drop 3.6 and lower we won't really need this test anymore.

asmeurer committed:
Use PYTHONIOENCODING=ascii:strict instead of LC_ALL=C in TEST_ASCII on Travis

This is because in Python 3.7, LC_ALL=C no longer results in Unicode errors
(PEP 538). Once we drop all Python versions lower than 3.7, we might consider
removing this test, as the environment it tests won't be very common.

asmeurer committed:
Don't pin the Python version in the Python 3 Travis optional dependency build

This will allow it to gracefully migrate from 3.6 to 3.7 once all the
dependencies support it (currently only tensorflow is missing). This is
similar to how the Sage build works.

asmeurer committed:
Use python: 3.6 for the Python 3 optional dependency build on Travis

We use the conda Python, not Travis's, so there is no point in using the
slower xenial/sudo: true container here.

asmeurer commented:
Actually, I've gone ahead and unpinned Python for the optional dependency build, so that it will automatically update once tensorflow gets a 3.7 package, same as with Sage. The Python version number on the build may be wrong, since that is the Travis Python version, which we aren't using (we use the conda Python). I've actually reverted it back to 3.6 because it doesn't require the (presumably) slower `sudo: true` container.

asmeurer committed:
Revert "Test not installing tensorflow in Python 3.7 on Travis"

This reverts commit 97ef658a81b0ae9ec13ba02766a8a571a61533a6.

asmeurer commented:
There are issues with symengine as well.  https://travis-ci.org/sympy/sympy/jobs/485558330

```
/home/travis/build/sympy/sympy/sympy/physics/mechanics/tests/test_functions.py 
Traceback (most recent call last):
  File "/home/travis/build/sympy/sympy/sympy/physics/mechanics/tests/test_functions.py", line 1, in <module>
    from sympy.core.backend import sin, cos, tan, pi, symbols, Matrix, zeros, S
  File "/home/travis/build/sympy/sympy/sympy/core/backend.py", line 6, in <module>
    from symengine import (Symbol, Integer, sympify, S,
  File "/home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/symengine/__init__.py", line 1, in <module>
    from .lib.symengine_wrapper import (Symbol, sympify as S, sympify,
ImportError: /home/travis/miniconda/envs/test-environment/lib/python3.7/site-packages/symengine/lib/symengine_wrapper.cpython-37m-x86_64-linux-gnu.so: undefined symbol: _ZN9SymEngine6expandERKNS_3RCPIKNS_5Bas
```

Hopefully they get fixed by the time tensorflow is fixed. CC @isuruf

isuruf commented:
Looking into it now

asmeurer committed:
Fix the Python 3.7 slow tests build on Travis

asmeurer commented:
Any objections?

asmeurer commented:
Going to merge this.
